controllers/remote/cluster_cache_test.go

```diff
@@ -0,0 +1,744 @@
+/*
+Copyright 2020 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package remote
```

Comment:
These tests are super hard to understand imo. I really dislike Ginkgo and wish this project would avoid it. But if we're going to use it, I would avoid using `It`s inside of helper functions. I think this code also uses nested `Context`s/`BeforeEach`s too much. It creates this situation where you have to constantly be scanning backwards in the code to understand how variables are scoped.

I think the best pattern is just to keep variables scoped to `It` functions and not relying on variable scoping/shadowing, and using helper functions to share code/reduce boilerplate. I know this is a big ask so I'm not going to make this a show-stopper or anything but this is what drives me crazy about Ginkgo.

Reply:
What about breaking tests in more files e.g.
- Cache and Predicates
- Reconcile
- Health cehcks

Reply:
I appreciate your concern and am also not entirely happy with how these turned out. I've fixed up the rest of the comments on this PR so will push that and will try and find some time to resist this and rework the tests to be a bit easier to follow. I think Fabrizio's suggestion of breaking the tests out into smaller files might make this more manageable so I'll start with that when I'm refactoring

Reply:
I've broken the tests into smaller files and reworked the reconciler test to use a table, does this make things clearer? Can have a go at the other tests as well if that makes sense

Reply:
It's fine I'll keep my ginkgo gripes to myself 😸